### PR TITLE
[TIR] Avoid unnecessary dtype escalation in loop splitting

### PR DESCRIPTION
**Pull request by zxybazh**
This PR introduces a type check to cast loop split decisions (sometimes given as `int64`) back to a smaller datatype when the loop variable's data type is smaller. This issue usually happens during reloading a trace from disk using JSON database and causes the failure of `CompactBufferAllocation` pass.

Unit test is also added.

### CODE REVIEW

**Comment by zxybazh** [comment]
Double checked and confirmed the only source of data type inconsistency comes from directly calling `Split` with given decisions. Therefore, I introduced a type check in `Split`'s factor parsing part. Ready for review. CC: @junrushao1994

File: src/tir/schedule/concrete_schedule.cc [hunk 1]
```diff
@@ -452,6 +452,9 @@ Array<LoopRV> ConcreteScheduleNode::Split(const LoopRV& loop_rv,
       if (is_const_int(factor) && !is_positive_const(factor)) {
         throw NonPositiveFactorError(state_->mod, factor.as<IntImmNode>()->value, i);
       }
+      if (factor.dtype().bits() > loop->extent.dtype().bits()) {
+        factor = IntImm(loop->extent.dtype(), factor.as<IntImmNode>()->value);
+      }
```
Comment (junrushao1994) [inline src/tir/schedule/concrete_schedule.cc:457]: A `factor` isn't necessary a constant integer, so casting it using `factor.as<IntImmNode>()` is probably not the best here. Instead please use:

```python
      if (factor.dtype().bits() > loop->extent.dtype().bits()) {
        factor = tir::cast(loop->extent.dtype(), factor);
      }
```

File: tests/python/unittest/test_meta_schedule_database.py [hunk 2]
```diff
@@ -19,17 +19,25 @@
...
 from typing import Callable
+import pytest
 
 import tvm
+from tvm.meta_schedule.arg_info import TensorInfo
+from tvm.meta_schedule.builder.builder import BuilderInput
+from tvm.meta_schedule.builder.local_builder import LocalBuilder
+from tvm.meta_schedule.database.database import TuningRecord
+from tvm.meta_schedule.runner.local_runner import LocalRunner
+from tvm.meta_schedule.runner.runner import RunnerInput
```
Comment (junrushao1994) [inline tests/python/unittest/test_meta_schedule_database.py:30]: this is very verbose and we have decided previously to use `ms.builder.LocalBuilder` instead to make the logic clearer. please see the second point in this PR description: https://github.com/apache/tvm/pull/11622

File: tests/python/unittest/test_meta_schedule_database.py [hunk 3]
```diff
@@ -68,9 +76,78 @@ def main(a: T.handle, b: T.handle, d: T.handle) -> None:  # pylint: disable=no-s
                 D[vi, vj] = T.max(C[vi, vj], 0.0)
 
 
-# fmt: on
-# pylint: enable=invalid-name,no-member,line-too-long,too-many-nested-blocks,no-self-argument
-
+@tvm.script.ir_module
+class MobileNetConv2d:
```
Comment (junrushao1994) [inline tests/python/unittest/test_meta_schedule_database.py:80]: first of all, this PR fixes a TIR issue, so it has nothing to do with MetaSchedule fundamentally; second, we should focus on "minimal" reproducible example while it's indeed not minimal enough.

alternatively, I would suggest
1) creating a unittest in `test_tir_schedule_split_fuse.py`
2) the IRModule contains a single loop
3) the schedule contains a single `split`, whose factor is `factors=[None, IntImm("int64", some-number)]`
4) check if the loop extent remains int32

**Comment by zxybazh** [comment]
Fixed and all issues addressed! Thanks for the feedbacks.

File: src/tir/schedule/concrete_schedule.cc [hunk 4]
```diff
@@ -452,6 +452,9 @@ Array<LoopRV> ConcreteScheduleNode::Split(const LoopRV& loop_rv,
       if (is_const_int(factor) && !is_positive_const(factor)) {
         throw NonPositiveFactorError(state_->mod, factor.as<IntImmNode>()->value, i);
       }
+      if (factor.dtype().bits() > loop->extent.dtype().bits()) {
+        factor = tir::Cast(loop->extent.dtype(), factor);
```
Comment (junrushao1994) [inline src/tir/schedule/concrete_schedule.cc:456]: use lower cased cast

Comment (zxybazh) [reply src/tir/schedule/concrete_schedule.cc]: Do you mean `tvm::cast` not `tir::cast` right? There seems to be no lowercased cast in tir namespace.

Comment (junrushao1994) [reply src/tir/schedule/concrete_schedule.cc]: right

**Review by junrushao1994 — APPROVED** [review]
LGTM!

**Comment by junrushao1994** [comment]
Thanks @zxybazh for looking into that!

**Pull request merged**